core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/DiskBasedChunkCache.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.chunkmanager.cache;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+
+import io.aiven.kafka.tieredstorage.chunkmanager.ChunkKey;
+import io.aiven.kafka.tieredstorage.chunkmanager.ChunkManager;
+
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.Weigher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+public class DiskBasedChunkCache extends ChunkCache<Path> {
+
+    private static final Logger log = LoggerFactory.getLogger(DiskBasedChunkCache.class);
+    static final String TEMP_CACHE_DIRECTORY = "temp";
+    static final String CACHE_DIRECTORY = "cache";
+
+    private DiskBasedChunkCacheConfig config;
+
+    public DiskBasedChunkCache(final ChunkManager chunkManager) {
+        super(chunkManager);
+    }
+
+    @Override
+    public InputStream cachedChunkToInputStream(final Path cachedChunk) {
+        try {
+            return Files.newInputStream(cachedChunk);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Writes the chunk to specified directory on disk.
+     * To be able to understand if the files are fully written to disk during a potential recovery process,
+     * this implementation initially writes them to temporary directory and then atomically copies them
+     * to the actual caching directory.
+     */
+    @Override
+    public Path cacheChunk(final ChunkKey chunkKey, final InputStream chunk) throws IOException {
+        final Path tempChunkPath = config.baseCachePath().resolve(TEMP_CACHE_DIRECTORY)
+                .resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(tempChunkPath.getParent())) {
+            Files.createDirectories(tempChunkPath.getParent());
+        }
+        final Path tempCached = writeToDisk(chunk, tempChunkPath);
+        log.info("Chunk file has been stored to temporary caching directory {}", tempCached);
+        final Path cachedChunkPath = config.baseCachePath()
+                .resolve(CACHE_DIRECTORY).resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(cachedChunkPath.getParent())) {
+            Files.createDirectories(cachedChunkPath.getParent());
+        }
+        final Path newPath = Files.move(tempCached, cachedChunkPath, ATOMIC_MOVE);
```

Comment:
Maybe if `move` fails, we should delete the temp file?

Reply:
For now I think we can do that but at the same time if the request will be retied by broker we will download the chunk again and this can be an infinite loop AFAIU. So maybe we should check maybe if we have the chunk in temp directory already before download? But we can address this later.

Reply:
That's what I was thinking about also. AFAIR there was some discussion about implementing some integrity checks. So let's postpone this and just delete for now I guess

Reply:
Yeah, that's reasonable

Reply:
Done